### DISCUSSION TITLE
Does openzfs for windows use caching in ram??

### DISCUSSION THREAD

**Discussion by Mokerak** [openzfsonwindows/openzfs#310] [Q&A] [answered]
When I've heard about ZFS I've heard allot it can use ram to cache stuff recently accessed to improve speed, is this true of the windows version?/ Can be enabled? 

This seems like something that it wouldn't do though for a windows port due to many issues

**Comment by Mokerak** [comment]
I'm bad at Github and very novice at linux so I'm not good at looking through configs

**Comment by lundman** [comment]
It is one of those cases, where people think they want RAM free, but that is a waste. You paid for them, and for the power to run the RAM. So the philosophy is generally, instead of having it sit free/empty, use it as a cache. And if someone wants to allocate some RAM, you can drop anything quickly that isn't being used. 

So in a perfect world, 99% of your RAM should be in use, and it always releases cached things when needed. That last part is hard, without a timemachine.

So yes, ZFS does use memory, because you want it to.

Now, platform specifics, we can't not react as quickly as needed, when Windows wants RAM, so we "pretend" you only have 50% of the memory, and stay under it that limit. Sort of like a VM. You can change this value, as required for your workload. Dedicated file server? go higher, heavy desktop use? Make it smaller.

At least, this is the goal. We are still tuning the memory cache part.

That cache is generally in "ARC". There is also L2ARC and SLOG, but they are caches for devices (like SSD/nvme).

**Comment by Mokerak** [comment]
Can I change how much ram it uses without recreating the whole thing?

Also, if you answer yes to this I probably wont want to.

If something else is using up say 75% of ram like a game, will ZFS only use 24.999%?? (Because that would be ideal)
Does it scale to the other systems uses of ram

**Comment by Mokerak** [comment]
Another question I have is how can I actually monitor how much ram it uses, since task manager and resource monitor can't see it

**Reply by EchterAgo** [reply]
The `kstat` command will give you a lot of information.

**Reply by lundman** [reply, marked as answer]
kstat will dump all, in the Unix format. 

Use registry to set the ARC limits:

`zfs_arc_max` from `\HKEY_LOCAL_MACHINE\SYSTEM\ControlSet001\Services\OpenZFS`

**Reply by Mokerak** [reply]
How do I read this? 

![image](https://github.com/openzfsonwindows/openzfs/assets/34557152/44480e72-0284-49a3-8aae-b21c04b84f87)

**Comment by lundman** [comment]
Quick read, go for "in_use" fields. Usually, arcstat.pl will read it and produce nicer output. Not sure how that runs on Windows.

**Comment by Mokerak** [comment]
> kstat will dump all, in the Unix format.
> 
> Use registry to set the ARC limits:
> 
> `zfs_arc_max` from `\HKEY_LOCAL_MACHINE\SYSTEM\ControlSet001\Services\OpenZFS`

I'm in the registry editor, what units does the zfs_arc_max use? would 8192 be 8GB, MB, KB?

**Reply by EchterAgo** [reply]
https://openzfs.github.io/openzfs-docs/Performance%20and%20Tuning/Module%20Parameters.html#zfs-arc-max

**Reply by sskras** [reply]
Apparently it's in **bytes**.

Thanks for an interesting thread. I see some folks are setting it to 1 GiB (1,073,741,824 bytes):
https://www.reddit.com/r/zfs/comments/vekogs/how_do_i_tune_zfs_for_low_ram_consumption/#t1_icqwtk3-comment-rtjson-content
